### PR TITLE
Add options dialog to Reflectometry GUI [PART 1]

### LINKED ISSUE #22051: No prompt to save changes on closing ISIS Reflectometry
### Expected behavior

When closing the interface after editing data, a prompt to save/discard/cancel should appear, if the changes have not been saved already.

### Actual behavior

The interface just closes, without saving the changes.

### Steps to reproduce the behavior

* Open ISIS Reflectometry.
* Edit some of the cells in the table
* Close the interface

### Platforms affected

OSX
Windows

### LINKED ISSUE #26533: Add Options and SlitCalculator for reflectometry GUI
The ISIS Reflectometry GUI used to have a menu option to open an options dialog and slit calculator. This should be re-added.

## Options dialog
This depends on PR #26835, which has done the work to add the Tools menu and an outline notification function and test for the presenter (see the TODOs in the code from that PR). You can apply this patch to add the Options dialog into the Tools menu and to connect the slot in the view:  [0001-Revert-Remove-Options-Dialog-from-the-menu.zip](https://github.com/mantidproject/mantid/files/3630680/0001-Revert-Remove-Options-Dialog-from-the-menu.zip)

Then the main remaining task is to implement the new Options dialog.

The dialog allows control of:
- the precision shown in the table
- whether warnings should be shown for:
  - table changes will be lost
  - you're about to process only a subset of a group
  - you're about to process all rows in the table

The warnings above should also be added (the only one currently implemented is to warn on losing table changes when starting autoprocessing, but we should also warn when closing the GUI)

~~There is potentially a bug where AutoProcess does not ask you if it is ok to overwrite a single group and row as that is effectively an "empty" table~~ - no longer a problem.

- [x] We can largely copy the layout from `DataProcessorOptionsDialog.ui` but the old dialog is too tied to the data processor to be  useful. 
- [x] The new dialog should be owned by `MainWindow`, with communication via presenters.
- [x] We'll need to load settings when the reflectometry interface is opened (i.e. when the dialog is constructed), similarly to `QDataProcessorWidget::loadSettings`, and save them when requested from the dialog.

### Slit Calculator
PR #26835

Note that this requires the current instrument, which should be synced across tabs, hence the dependency on #26536. The existing slit calculator dialog can be used so we just need to add a menu option and notifications to open it. Note that it also needs to be updated after the instrument changes.

### PR DESCRIPTION
**Pull request by giovannidisiena**
**Description of work.**
This work is a combination of several PRs (which have all been reviewed and merged into this branch) to add an Options dialog and related settings to the ISIS reflectometry interface.

**To test:**
Complete a code review. Full manual testing instructions:
- Check that the dialog is displayed when requested
- Check that **warn discard changes** is displayed when checked and expected (e.g. editing and closing a batch, loading a new batch on top of existing changes, etc) and the negative case (e.g. warning should not be displayed if a batch is first saved, then closed/overwritten). The following should now all be correct:

> Works well in these scenarios:

>- Warning appears when loading a batch after table has changed, and respects the setting.
>- Warning appears when starting autoprocess, and respects the setting.
>- Warning does not appear after pausing and restarting autoprocessing even if the table has >changed, because the table does not get overwritten unless the search criteria have changed. The >warning does appear if the search criteria have changed.
>- Warning appears after changing settings on Event/Experiment/Instrument tabs.
>
>However, seems to be incorrect in these scenarios:
>
>- Warning does not appear after editing the Save tab.
>- Warning when starting autoprocess does not respect the setting.
>- Batch is incorrectly marked as unsaved after processing a row (when state has changed but no content in the table has actually changed)
>- Warning incorrectly appears when I `Restore Defaults` in Experiment and Instrument tabs if nothing has changed. I don't think it's necessary to have a warning here so we can just remove this.

- Check that **warn process all/partial** is displayed when checked and expected (e.g. no rows selected, all groups with content selected, all rows selected, etc). The following should all work:

> A few problems showed up in functional testing. I'm working with the INTER_NR_test.json table from the ISIS sample data.

>- The warning correctly appears when partially processing a group but does not respect the setting when I turn it off.
>- If I select all of the rows within a group, I get a spurious warning that the whole table will be processed - if I turn it off, the whole table does indeed get processed, when it should not - this is a regression. (I also noticed that if I re-load the table, the progress bar does not reset, but this is an existing issue so not related to this PR.)
>- If I select a group (but not its contained rows) and process, then I get a warning that the group will not be fully processed, when actually it will.
>- If I select a group, including its contained rows, and then process, then I get a crash.

-  Check that the **table precision** is displayed as expected (e.g. default value if unchecked, user-specified value if checked, default value if unchecked again, etc).
- Check that all tests pass.

Fixes #26533 and fixes #22051 

*This does not require release notes* because **they will be added in a subsequent PR**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/inc/MantidQtWidgets/Common/QSettingsHelper.h
Comment (ConorMFinn) [inline qt/widgets/common/inc/MantidQtWidgets/Common/QSettingsHelper.h]: `settingTypeName` looks like it could be defined here, avoiding creating the string object if it does not end in `"/type"`

File: qt/widgets/common/inc/MantidQtWidgets/Common/QSettingsHelper.h
Comment (ConorMFinn) [inline qt/widgets/common/inc/MantidQtWidgets/Common/QSettingsHelper.h]:
```suggestion
 *  as QVariant types are not properly encoded in ini files
```

File: qt/widgets/common/inc/MantidQtWidgets/Common/QSettingsHelper.h
Comment (ConorMFinn) [inline qt/widgets/common/inc/MantidQtWidgets/Common/QSettingsHelper.h]:
```suggestion
#endif /* MANTID_QSETTINGSHELPER_H */
```

File: qt/scientific_interfaces/ISISReflectometry/GUI/Options/OptionsDialogModel.cpp
Comment (ConorMFinn) [inline qt/scientific_interfaces/ISISReflectometry/GUI/Options/OptionsDialogModel.cpp]: Personally, I'm not a fan of using out arguments like this. Though I understand the reasoning as you have two maps to work with.

**Review by cailafinn — CHANGES_REQUESTED** [review]
Looks good so far, just a couple of typos and comments.

**Review by cailafinn — DISMISSED** [review]
This looks good to me now. Good work :+1:

**Comment by gemmaguest** [comment]
Is it worth tagging this as Not Ready until the other PRs are merged?

**Comment by giovannidisiena** [comment]
> This looks good to me now. Good work 👍

Thanks @ConorMFinn, merging #27656 dismissed your review so may be worth waiting until the end to have another once-over

**Comment by gemmaguest** [comment]
@giovannidisiena Could you ensure that the full manual testing instructions are ready for when part 6 is merged?

**Comment by gemmaguest** [comment]
This bug seems to have got back in: Batch is incorrectly marked as unsaved after processing a row (when state has changed but no content in the table has actually changed)

**Comment by giovannidisiena** [comment]
> This bug seems to have got back in: Batch is incorrectly marked as unsaved after processing a row (when state has changed but no content in the table has actually changed)

Think it got back in via a rebase, should be gone now!

**Comment by giovannidisiena** [comment]
Reminder: this PR will likely need to updating due to #28200, #28320, #28279 and #27061

**Comment by gemmaguest** [comment]
@giovannidisiena Have you seen the conflicts on this? Is it ready, otherwise?

**Comment by giovannidisiena** [comment]
@gemmaguest yes but for some reason not authorised to resolve them... and not quite, need to do a little but of updating as per the maintenance PRs above.

**Comment by giovannidisiena** [comment]
@gemmaguest this is now up to date with the changes introduced by maintenance issues.

**Review by gemmaguest — CHANGES_REQUESTED** [review]
The reflectometry side of things now all looks good and everything is working as expected 🎉 

We should also test the Indirect interface(s) that use the settings helper. Do you know how/what to test here? I guess it is the `Interfaces->Indirect->Settings` dialog. Unfortunately if I change the settings and close and re-open this dialog it doesn't remember the changes, which looks like a regression.

**Comment by giovannidisiena** [comment]
@gemmaguest I just tested this on `5.0` and it seems that if you re-open the settings dialog soon after closing then the settings are 'remembered', otherwise given an amount of time they will be reset. I introduced the first use of `QSettingsHelper` into the indirect code in this PR so it doesn't seem like a regression to me, maybe more an issue of the underlying code which is handling those settings once they are returned?

**Comment by giovannidisiena** [comment]
@gemmaguest the other reason for this may the new format I used to save the settings due to .ini file encoding, which means that you probably need to clear the existing settings `plot-error-bars-external` and `restrict-input-by-name`...

**Comment by giovannidisiena** [comment]
@gemmaguest sorry for the spam of ideas, to me it looks like this code is overriding any changes to the `restrict input data by name` option...

https://github.com/mantidproject/mantid/blob/87284a6b7b914d2fa2518a1492e80bda129dc4d1/qt/scientific_interfaces/Indirect/IndirectSettingsPresenter.cpp#L35-#L42

**Comment by giovannidisiena** [comment]
@gemmaguest sorry for the spam of ideas, to me it looks like this code is overriding any changes to the `restrict input data by name` option:

https://github.com/mantidproject/mantid/blob/87284a6b7b914d2fa2518a1492e80bda129dc4d1/qt/scientific_interfaces/Indirect/IndirectSettingsPresenter.cpp#L34-#L42

**Comment by gemmaguest** [comment]
> 
> 
> @gemmaguest sorry for the spam of ideas (and GH messing up), to me it looks like this code is overriding any changes to the `restrict input data by name` option:
> 
> [/qt/scientific_interfaces/Indirect/IndirectSettingsPresenter.cpp@`87284a6`#L34-#L42](https://github.com/mantidproject/mantid/blob/87284a6b7b914d2fa2518a1492e80bda129dc4d1/qt/scientific_interfaces/Indirect/IndirectSettingsPresenter.cpp#L34-#L42)

That's good to know - looks like there's also a minor bug with this default setting where this bit of code sets the default to true in the file when you open the dialog, but does not update the view so it remains unchecked... But if you then close and re-open the dialog then it is shown as checked as per the file...

However, testing with the other setting it works as I expect in master - whatever is in the file is reflected on the GUI when you open it. But I can't get it to work on this PR's branch, regardless of whether I remove the old format settings from the file or not. The new format setting is written correctly to the file, but when I re-open the dialog it does not pick it up. Playing around with the file contents, it looks like what is actually happening is that it is still reading from the old setting format even though it is writing to the new setting format??

**Comment by giovannidisiena** [comment]
@gemmaguest that's strange, I'm not sure how/why it doesn't read from the new format if the old one is removed as the group name hasn't changed. In any case, this part is slightly outside the scope of this PR so could be left to someone who understands the .ini file/indirect interface better if they decide they want to use it...

**Comment by gemmaguest** [comment]
The Indirect settings are working again :+1:

Could you add release notes? (They seem to have disappeared in the last force-push because they were in Muon for some reason!)

**Review by gemmaguest — APPROVED** [review]
Thanks for adding the release notes. This all looks good now :+1:

**Pull request merged**
